// stylelint-selector-pseudo-class-lvhfa/cjs/index.js
'use strict';

var stylelint = require('stylelint');
var without = require('lodash.without');
var difference = require('lodash.difference');
var selectorParser = require('postcss-selector-parser');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var stylelint__default = /*#__PURE__*/_interopDefaultLegacy(stylelint);
var without__default = /*#__PURE__*/_interopDefaultLegacy(without);
var difference__default = /*#__PURE__*/_interopDefaultLegacy(difference);
var selectorParser__default = /*#__PURE__*/_interopDefaultLegacy(selectorParser);

const ruleName = 'plugin/selector-pseudo-class-lvhfa';
const messages = stylelint__default["default"].utils.ruleMessages(ruleName, {
  expected: 'Expected pseudo class selectors to follow LVHFA order.'
});
const correctOrder = [':link', ':visited', ':hover', ':focus', ':active'];

/**
 * @type {stylelint.RuleBase}
 */
function ruleFunction( /** @type boolean */bool) {
  return (cssRoot, result) => {
    const validOptions = stylelint__default["default"].utils.validateOptions(result, ruleName, {
      actual: bool
    });
    if (!validOptions) {
      return;
    }
    cssRoot.walkRules(rule => {
      selectorParser__default["default"](selector => {
        /** @type {string[]} */
        let inputOrder = [];
        selector.walkPseudos(pseudo => {
          if (correctOrder.indexOf(pseudo.value) !== -1) {
            inputOrder.push(pseudo.value);
          }
        });
        inputOrder = [...new Set(inputOrder)];
        const finalOrder = without__default["default"](correctOrder, ...difference__default["default"](correctOrder, inputOrder));
        const finalResult = finalOrder.every((pseudo, index) => pseudo === inputOrder[index]);
        if (!finalResult) {
          stylelint__default["default"].utils.report({
            ruleName: ruleName,
            result: result,
            node: rule,
            message: messages.expected
          });
        }
      }).processSync(rule.selector);
    });
  };
}

// @ts-ignore
const plugin = stylelint__default["default"].createPlugin(ruleName, ruleFunction);
var index = {
  ...plugin,
  messages
};

module.exports = index;
